sdk/search/Azure.Search.Documents/src/Batching/Publisher{T}.cs

```diff
@@ -0,0 +1,505 @@
+
+            // Send the request
+            bool throttled = await OnSubmitBatchAsync(batch, cancellationToken).ConfigureAwait(false);
+
+            // Update whether or not the request was throttled to update our
+            // retry delay
+            _manualRetries.Update(throttled);
+        }
+
+        /// <summary>
+        /// Send indexing actions to be processed by the service.
+        /// </summary>
+        /// <param name="batch">The batch of actions to submit.</param>
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns>Whether the submission was throttled.</returns>
+        protected abstract Task<bool> OnSubmitBatchAsync(IList<PublisherAction<T>> batch, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Enqueue an action to be retried.
+        /// </summary>
+        /// <param name="action">The action to be retried.</param>
+        /// <param name="skipIncrement">
+        /// Whether we should skip incrementing the retry attempts because the
+        /// failure wasn't related to the action.
+        /// </param>
+        /// <returns>
+        /// A value indicating whether the action is retriable.
+        /// </returns>
+        protected bool EnqueueRetry(
+            PublisherAction<T> action,
+            bool skipIncrement = false)
+        {
+            bool retriable = skipIncrement || action.RetryAttempts++ < RetryCount;
+            if (retriable)
+            {
+                _retry.Enqueue(action);
+            }
+            return retriable;
+        }
+        #endregion Publishing
+
+        #region Timer
+        /// <summary>
+        /// Start the timer if it's not already running.
+        /// </summary>
+        private void StartTimer()
+        {
+            if (AutoFlushInterval != null)
+            {
+                int intervalInMs = (int)AutoFlushInterval.Value.TotalMilliseconds;
+                if (_timer == null)
+                {
+                    _timer = new Timer(
+                        OnTimerElapsed,
+                        state: null,
+                        dueTime: intervalInMs,
+                        period: Timeout.Infinite);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Stop the timer.
+        /// </summary>
+        private void StopTimer()
+        {
+            if (_timer != null)
+            {
+                _timer.Dispose();
+                _timer = null;
+            }
+        }
+
+        /// <summary>
+        /// Trigger a flush when the timer elapses.
+        /// </summary>
+        /// <param name="timerState"></param>
+        private void OnTimerElapsed(object timerState)
+        {
+            // We'll do this synchronously since it's an unbounded channel and
+            // we can just spin if we need to.
+            _channel.Writer.WriteInternal(
```

Comment:
Would this crash the process  if it throws?

Reply:
The comment about spinning is out of date.  And good suggestion about exceptions.  I'd rather not have the timer fire than blow everything up.